nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/smoother.hpp

```diff
@@ -0,0 +1,377 @@
+            auto & current = path_optim[i];
+            Eigen::Vector2d tangent_dir = tangentDir<double>(
+              prelast.block<2, 1>(0, 0),
+              last.block<2, 1>(0, 0),
+              current.block<2, 1>(0, 0),
+              prelast[2] * last[2]);
+
+            last_dir =
+              tangent_dir.dot((current - last).block<2, 1>(0, 0) * last[2]) >= 0 ?
+              tangent_dir :
+              -tangent_dir;
+            last_dir.normalize();
+          } else {
+            last_dir = end_dir;
+          }
+          double last_angle = atan2(last_dir[1], last_dir[0]);
+
+          // Interpolate poses between prelast and last
+          int interp_cnt = (last_i - prelast_i) * params.output_upsampling_factor - 1;
+          if (interp_cnt > 0) {
+            Eigen::Vector2d lastp = last.block<2, 1>(0, 0);
+            Eigen::Vector2d prelastp = prelast.block<2, 1>(0, 0);
+            double dist = (lastp - prelastp).norm();
+            Eigen::Vector2d p1 = prelastp + prelast_dir * dist * 0.4 * prelast[2];
+            Eigen::Vector2d p2 = lastp - last_dir * dist * 0.4 * prelast[2];
+            for (int j = 1; j <= interp_cnt; j++) {
+              double interp = j / static_cast<double>(interp_cnt + 1);
+              Eigen::Vector2d p = cubicBezier(prelastp, p1, p2, lastp, interp);
+              path.emplace_back(p[0], p[1], 0.0);
+            }
+          }
+          path.emplace_back(last[0], last[1], last_angle);
+
+          // Assign orientations to interpolated points
+          for (size_t j = path.size() - 1 - interp_cnt; j < path.size() - 1; j++) {
+            Eigen::Vector2d tangent_dir = tangentDir<double>(
+              path[j - 1].block<2, 1>(0, 0),
+              path[j].block<2, 1>(0, 0),
+              path[j + 1].block<2, 1>(0, 0),
+              1.0);
+            tangent_dir =
+              tangent_dir.dot((path[j + 1] - path[j]).block<2, 1>(0, 0) * prelast[2]) >= 0 ?
+              tangent_dir :
+              -tangent_dir;
+            path[j][2] = atan2(tangent_dir[1], tangent_dir[0]);
+          }
+
+          prelast_dir = last_dir;
+        }
+        prelast_i = last_i;
+        last_i = i;
+      }
+    }
+  }
+
+  /*
+    Piecewise cubic bezier curve as defined by Adobe in Postscript
+    The two end points are p0 and p3
+    Their associated control points are p1 and p2
+  */
+  static Eigen::Vector2d cubicBezier(
+    Eigen::Vector2d & p0, Eigen::Vector2d & p1,
+    Eigen::Vector2d & p2, Eigen::Vector2d & p3, double mu)
+  {
+    Eigen::Vector2d a, b, c, p;
+
+    c[0] = 3 * (p1[0] - p0[0]);
+    c[1] = 3 * (p1[1] - p0[1]);
+    b[0] = 3 * (p2[0] - p1[0]) - c[0];
+    b[1] = 3 * (p2[1] - p1[1]) - c[1];
+    a[0] = p3[0] - p0[0] - c[0] - b[0];
+    a[1] = p3[1] - p0[1] - c[1] - b[1];
+
+    p[0] = a[0] * mu * mu * mu + b[0] * mu * mu + c[0] * mu + p0[0];
+    p[1] = a[1] * mu * mu * mu + b[1] * mu * mu + c[1] * mu + p0[1];
+
+    return p;
+  }
+
+  bool _debug;
```

Comment:
Change to trailing underscore so consistent within the package across all files

Reply:
Actually, you can just make the small number of variable_'s into _variables's, it'll be easier for you than to change all of these all over the place for just 3